packages/components/src/components/SquiggleViewer/utils.ts

```diff
@@ -66,72 +44,65 @@ export function getChildrenValues(value: SqValue): SqValue[] {
 export function useGetSubvalueByPath() {
   const { itemStore } = useViewerContext();
 
-  return (value: SqValue, path: SqValuePath): SqValue | undefined => {
-    const { context } = value;
-    if (!context) {
-      return;
-    }
-    if (context.path.root !== path.root) {
-      return;
-    }
-    if (context.path.items.length > path.items.length) {
+  return (
+    topValue: SqValue,
+    subValuePath: SqValuePath
+  ): SqValue | undefined => {
+    const { context } = topValue;
+
+    if (!context || !subValuePath.contains(context.path)) {
       return;
     }
 
-    for (let i = 0; i < path.items.length; i++) {
-      if (i < context.path.items.length) {
-        // check that `path` is a subpath of `context.path`
-        if (!isEqual(context.path.items[i], path.items[i])) {
+    let currentValue = topValue;
+    const subValuePaths = subValuePath.allSqValuePathSubsets({
+      includeRoot: false,
+    });
+
+    for (const subValuePath of subValuePaths) {
+      const pathEdge = subValuePath.lastItem()!; // We know it's not empty, because includeRoot is false.
```

Comment:
Some time ago I converted this function to traverse the path items manually, instead of pre-generating all prefixes, because of a habit to fight all `O(N^2)` that I encounter; now you converted it back...

I guess because you disliked C-style `for(;;)` loop and the way I constructed `calculatorPath`? Oh well.

The real wart on this function, though, is that it depends on `itemStore`. Otherwise it would belong to `SqValue` as a method.

Maybe we can still move it there?

```typescript
class SqAbstractValue {
  ...
  getSubvalueByPath(
    path: SqValuePath,
    traverseCalculatorEdge: (path: SqValuePath) => SqValueResult
  ) {
    ...
  }
```

You could even try to split the implementation into specific `SqValue` classes by implementing another function, `getPathEdge(edge: SqValuePathEdge)` (but I'm not sure how to handle calculators with this)